worker/pubsub/remoteserver.go

```diff
@@ -0,0 +1,336 @@
+	Hub    *pubsub.StructuredHub
+	Origin string
+	Target string
+	Clock  clock.Clock
+
+	// APIInfo is initially populated with the addresses of the target machine.
+	APIInfo    *api.Info
+	APIOpen    func(*api.Info, api.DialOpts) (api.Connection, error)
+	WriterOpen func(api.Connection) (pubsubapi.MessageWriter, error)
+}
+
+// NewRemoteServer creates a new RemoteServer that will connect to the remote
+// apiserver and pass on messages to the pubsub endpoint of that apiserver.
+func NewRemoteServer(config RemoteServerConfig) (RemoteServer, error) {
+	closed := make(chan struct{})
+	close(closed)
+	remote := &remoteServer{
+		origin:     config.Origin,
+		target:     config.Target,
+		info:       config.APIInfo,
+		openAPI:    config.APIOpen,
+		openWriter: config.WriterOpen,
+		hub:        config.Hub,
+		clock:      config.Clock,
+		pending:    deque.New(),
+		closed:     closed,
+		data:       make(chan struct{}),
+	}
+	unsub, err := remote.hub.Subscribe(forwarder.ConnectedTopic, remote.onForwarderConnection)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	remote.unsubscribe = unsub
+	go func() {
+		defer remote.tomb.Done()
+		remote.tomb.Kill(remote.loop())
+	}()
+	return remote, nil
+}
+
+func (r *remoteServer) onForwarderConnection(topic string, details forwarder.OriginTarget, err error) {
+	if err != nil {
+		// This should never happen.
+		logger.Errorf("subscriber callback error: %v", err)
+		return
+	}
+	if details.Target == r.origin && details.Origin == r.target {
+		// If we have just been connected to by the apiserver that we are
+		// trying to connect to, interrupt any waiting we may be doing and try
+		// again.
+		r.interruptConnecting()
+	}
+}
+
+// UpdateAddresses will update the addresses held for the target API server.
+// If we are currently trying to connect to the target, interrupt it so we
+// can try again with the new addresses.
+func (r *remoteServer) UpdateAddresses(addresses []string) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	if r.connection == nil && r.stopConnecting != nil {
+		// We are probably trying to reconnect, so interrupt that so we don't
+		// get a race between setting addresses and trying to read them to
+		// connect. Note that we don't call the interruptConnecting method
+		// here because that method also tries to lock the mutex.
+		logger.Debugf("interrupting connecting due to new addresses: %v", addresses)
+		close(r.stopConnecting)
+		r.stopConnecting = nil
+	}
+	r.info.Addrs = addresses
+}
+
+// Publish queues up the message if and only if we have an active connection to
+// the target apiserver.
+func (r *remoteServer) Publish(message *params.PubSubMessage) {
+	select {
+	case <-r.tomb.Dying():
+		logger.Tracef("dying, don't send %q", message.Topic)
+	default:
+		r.mutex.Lock()
+		defer r.mutex.Unlock()
+		// Only queue the message up if we are currently connected.
+		if r.connection != nil {
+			logger.Tracef("queue up topic %q", message.Topic)
+			r.pending.PushBack(message)
```

Comment:
If the server hangs up we discard everything that is pending, and the connection is set to nil, so items stop getting added. We have an implicit guarantee that if the servers are connected, they get all the messages.